build-system/babel-plugins/babel-plugin-transform-log-methods/index.js

```diff
@@ -0,0 +1,288 @@
+  ) {
+    return null;
+  }
+  // object property is transformable method.
+  const {property} = callee;
+  if (!t.isIdentifier(property) || !(property.name in transformableMethods)) {
+    return null;
+  }
+  return transformableMethods[property.name];
+}
+
+module.exports = function({types: t}) {
+  return {
+    /** Resolves plugin config. */
+    pre() {
+      const {
+        // Configurable to isolate test output.
+        messagesPath = defaultMessagesPath,
+
+        // Temporary option to not replace call arguments, but still output the
+        // table to keep build code and infra independent from rollout.
+        replaceCallArguments = true,
+      } = this.opts;
+
+      this.messagesPath = relativeToRoot(messagesPath);
+      this.replaceCallArguments = replaceCallArguments;
+    },
+    visitor: {
+      /** Message table I/O. */
+      Program: {
+        enter() {
+          this.messages = getMessages(this.messagesPath);
+          this.nextMessageId = Object.keys(this.messages).length;
+          this.getMessageId = () => this.nextMessageId++;
+        },
+        exit() {
+          writeMessages(this.messagesPath, this.messages);
+        },
+      },
+      /**
+       * Visits call expressions for known log assertion/error methods.
+       * @param {*} path babel.path
+       */
+      CallExpression({node}) {
+        const meta = getTransformableCallExpressionMeta(t, node);
+        if (!meta) {
+          return;
+        }
+
+        const {indirectable, messageArgPos} = meta;
+        if (!indirectable) {
+          return;
+        }
+
+        const messageArg = node.arguments[messageArgPos];
+
+        if (!messageArg) {
+          return;
+        }
+
+        // Construct a printf template from the argument set. There could be
+        // non-string types among string literals in variadic calls, so the
+        // template includes them as interpolated arguments.
+        const templateArgs = [];
+        const message = buildMessage(t, messageArg, templateArgs);
+
+        // Bounce when indirection increases minified size (±1 byte). Also
+        // catches the the case where the top-level message is variable (ie.
+        // when its template is exactly '%s'), in which indirection would be
+        // pointless.
+        if (message.length <= roughMinifiedExpansionLength) {
+          return;
+        }
+
+        const shortMessageId = getOrCreateShortMessageId(
+          this.messages,
+          message,
+          this.getMessageId
+        );
+
+        if (!this.replaceCallArguments) {
+          return;
+        }
+
+        // We care only about message arguments beyond the first (ie. the body):
+        // all previous method arguments (ie. the head) should stay in place.
+        // The body is nested into an array expression, after the template id.
+        const headArgs = node.arguments.slice(0, Math.max(0, messageArgPos));
```

Comment:
Why the max?

Reply:
Leftover from doing a subtraction for some reason, before review. Thanks for catching.